aptos-move/framework/aptos-stdlib/sources/data_structures/btree_map.move

```diff
@@ -0,0 +1,1201 @@
+            // In this case, the key is greater than all keys in the tree.
+            leaf = self.max_leaf_index;
+            let current = self.nodes.borrow(leaf).parent;
+            while (current != NULL_INDEX) {
+                let current_node = self.nodes.borrow_mut(current);
+                let last_index = current_node.children.length() - 1;
+                let last_element = current_node.children.borrow_mut(last_index);
+                last_element.max_key = key;
+                current = current_node.parent;
+            }
+        };
+
+        self.add_at(leaf, new_leaf_child(key, value));
+    }
+
+    /// If the key doesn't exist in the tree, inserts the key/value, and returns none.
+    /// Otherwise updates the value under the given key, and returns the old value.
+    public fun upsert<K: drop + copy + store, V: store>(self: &mut BTreeMap<K, V>, key: K, value: V): Option<V> {
+        if (self.inner_max_degree == 0 || self.leaf_max_degree == 0) {
+            self.init_max_degrees(&key, &value);
+        };
+
+        let iter = self.find(key);
+        if (is_end_iter(self, &iter)) {
+            self.add(key, value);
+            return option::none()
+        } else {
+            let node = self.nodes.borrow_mut(iter.node_index);
+            let children = &mut node.children;
+
+            // Field swap doesn't compile.
+            // let child = vector::borrow_mut(children, iter.child_index);
+            // assert!(child.max_key == key, E_INTERNAL);
+            // let old = child.value;
+            // child.value = value;
+            // option::some(old)
+
+            let Child::Leaf {
+                max_key: old_max_key,
+                value: old_value,
+            } = children.replace(iter.child_index, Child::Leaf { max_key: key, value: value });
+            assert!(old_max_key == key, E_INTERNAL);
+            option::some(old_value)
+        }
+    }
+
+    /// Removes the entry from BTreeMap and returns the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun remove<K: drop + copy + store, V: store>(self: &mut BTreeMap<K, V>, key: K): V {
+        let iter = self.find(key);
+        assert!(!is_end_iter(self, &iter), E_INTERNAL);
+
+        let Child::Leaf {
+            value,
+            max_key: _,
+        } = self.remove_at(iter.node_index, key);
+
+        value
+    }
+
+    ///////////////
+    // Accessors //
+    ///////////////
+
+    // Returns true iff the iterator is a begin iterator.
+    public fun is_begin_iter<K: store, V: store>(tree: &BTreeMap<K, V>, iter: &Iterator<K>): bool {
+        if (iter is Iterator::End<K>) {
+            tree.is_empty()
+        } else {
+            (iter.node_index == tree.min_leaf_index && iter.child_index == 0)
+        }
+    }
+
+    // Returns true iff the iterator is an end iterator.
+    public fun is_end_iter<K: store, V: store>(_tree: &BTreeMap<K, V>, iter: &Iterator<K>): bool {
```

Comment:
this feels more like inline func than actual func?

Reply:
cannot be done publicly, as enum variants are not public. also some implementation might actually need to do something here.